File: src/include/asset/StoredAsset.php
<?php

namespace asset;

use creator\Creator;
use DateTime;
use Exception;
use log\Log;
use database\Database;

/**
 * The main asset class.
 * It represents one PBR material, 3D model or other asset.
 */
class StoredAsset extends Asset
{
	public function __construct(
		?int $id,
		?string $creatorGivenId,
		string $title,
		string $url,
		DateTime $date,
		AssetType $type,
		Creator $creator,
		array $tags = [],

		public StoredAssetStatus $status = StoredAssetStatus::ACTIVE,
		public ?DateTime $lastSuccessfulValidation = NULL,
	) {
		parent::__construct(
			id: $id,
			creatorGivenId: $creatorGivenId,
			title: $title,
			url: $url,
			date: $date,
			type: $type,
			creator: $creator,
			tags: $tags,
		);
	}

	public function getThumbnailUrl(int $size, string $extension, ?string $backgroundColor): string
	{
		$variation = strtoupper(implode("-", array_filter([$size, $extension, $backgroundColor])));
		$extension = strtolower($extension);
		$id = $this->id;
		return "/thumbnail/$variation/$id.$extension";
	}

	public function writeToDatabase()
	{

		if ($this->id) {
			Log::write("Updating asset", $this);

			// Save base asset
			$sql = "UPDATE Asset SET creatorGivenId=?, title=?,state=?,url=?,date=?,typeId=?,creatorId=?,lastSuccessfulValidation=? WHERE id = ?";
			$parameters = [$this->creatorGivenId, $this->title, $this->status->value, $this->url, $this->date, $this->type->value, $this->creator->value, $this->lastSuccessfulValidation, $this->id];
			Database::runQuery($sql, $parameters);

			// Tags
			Database::runQuery("DELETE FROM Tag WHERE id = ?", [$this->id]);
			foreach ($this->tags as $tag) {
				$sql = "INSERT INTO Tag (id,tag) VALUES (?,?);";
				$parameters = [$this->id, $tag];
				Database::runQuery($sql, $parameters);
			}
		} else {
			Log::write("Inserting new asset", $this);

			// Base Asset
			$sql = "INSERT INTO Asset (id,creatorGivenId, state, title, url, date, clicks, typeId, creatorId) VALUES (NULL, ?, ?, ?, ?, ?, ?, ?, ?);";
			$parameters = [$this->creatorGivenId, $this->status->value, $this->title, $this->url, $this->date, 0, $this->type->value, $this->creator->value];
			Database::runQuery($sql, $parameters);

			// Add the missing id to the asset object
			$this->id = Database::runQuery("SELECT id FROM Asset WHERE url = ?;", [$this->url])->fetchArray()['id'];
			if (!$this->id) {
				throw new Exception("Failed to retrieve ID of newly inserted asset with URL " . $this->url);
			}

			// Tags
			foreach ($this->tags as $tag) {
				$sql = "INSERT INTO Tag (id,tag) VALUES (?,?);";
				$parameters = [$this->id, $tag];
				Database::runQuery($sql, $parameters);
			}
		}
	}
}

File: app/asset/stored_asset.py
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional, TYPE_CHECKING

from app.asset.asset import Asset
from app.asset.stored_asset_status import StoredAssetStatus

if TYPE_CHECKING:
    from app.thumbnail.thumbnail_format import ThumbnailFormat


@dataclass
class StoredAsset(Asset):
    date: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    status: StoredAssetStatus = StoredAssetStatus.ACTIVE
    last_successful_validation: Optional[datetime] = None
    clicks: int = 0

    def __post_init__(self) -> None:
        now = datetime.now(timezone.utc)
        if self.date.tzinfo is None:
            self.date = self.date.replace(tzinfo=timezone.utc)
        if self.date > now:
            self.date = now

    def get_thumbnail_url(self, fmt: "ThumbnailFormat", full_url: bool = False, base_url: str = "") -> str:
        variation = fmt.value
        extension = fmt.get_extension().lower()
        url = base_url.rstrip("/") if full_url else ""
        url += f"/thumbnail/{variation}/{self.id}.{extension}"
        return url

    def api_representation(self, fmt: "ThumbnailFormat", base_url: str = "") -> dict:
        return {
            "id": self.id,
            "creatorGivenId": self.creator_given_id,
            "title": self.title,
            "url": self.url,
            "date": self.date.isoformat(),
            "type": self.type.slug(),
            "creator": self.creator.slug(),
            "tags": self.tags,
            "thumbnail": self.get_thumbnail_url(fmt, full_url=True, base_url=base_url),
        }

    def write_to_database(self) -> None:
        from app.database.database import Database
        from app.log.log import Log
        from app.log.log_level import LogLevel

        date_str = self.date.strftime("%Y-%m-%d %H:%M:%S")
        lsv_str = self.last_successful_validation.strftime("%Y-%m-%d %H:%M:%S") if self.last_successful_validation else None

        if self.id:
            Log.write("Updating asset", self.title, LogLevel.DEBUG)
            sql = (
                "UPDATE Asset SET creatorGivenId=?, title=?, state=?, url=?, date=?, "
                "typeId=?, creatorId=?, lastSuccessfulValidation=? WHERE id = ?"
            )
            params = [
                self.creator_given_id, self.title, self.status.value, self.url,
                date_str, self.type.value, self.creator.value, lsv_str, self.id,
            ]
            Database.run_query(sql, params)
            Database.run_query("DELETE FROM Tag WHERE id = ?", [self.id])
            for tag in self.tags:
                Database.run_query("INSERT INTO Tag (id, tag) VALUES (?, ?)", [self.id, tag])
        else:
            Log.write("Inserting new asset", self.title, LogLevel.DEBUG)
            sql = (
                "INSERT INTO Asset (id, creatorGivenId, state, title, url, date, clicks, typeId, creatorId) "
                "VALUES (NULL, ?, ?, ?, ?, ?, ?, ?, ?)"
            )
            params = [
                self.creator_given_id, self.status.value, self.title, self.url,
                date_str, 0, self.type.value, self.creator.value,
            ]
            cursor = Database.run_query(sql, params)
            self.id = cursor.lastrowid
            for tag in self.tags:
                Database.run_query("INSERT INTO Tag (id, tag) VALUES (?, ?)", [self.id, tag])
